virtual-adapter/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+#![no_main]
+
+wit_bindgen::generate!({
+    path: "../wit",
+    world: "virtual-adapter"
+});
+
+use crate::exports::wasi::cli_base::environment::Environment;
+
+struct VirtAdapter;
+
+export_virtual_adapter!(VirtAdapter);
+
+#[repr(C)]
+pub struct Env {
+    /// Whether to fallback to the host env
+    /// [byte 0]
+    host_fallback: bool,
+    /// Whether we are providing an allow list or a deny list
+    /// on the fallback lookups
+    /// [byte 1]
+    host_fallback_allow: bool,
+    /// How many host fields are defined in the data pointer
+    /// [byte 4]
+    host_field_cnt: u32,
+    /// Host many host fields are defined to be allow or deny
+    /// (these are concatenated at the end of the data with empty values)
+    /// [byte 8]
+    host_allow_or_deny_cnt: u32,
+    /// Byte data of u32 byte len followed by string bytes
+    /// up to the lengths previously provided.
+    /// [byte 12]
+    host_field_data: *const u8,
+}
+
+#[no_mangle]
+pub static mut env: Env = Env {
+    host_fallback: true,
+    host_fallback_allow: false,
+    host_field_cnt: 0,
+    host_allow_or_deny_cnt: 0,
+    host_field_data: 0 as *const u8,
+};
+
+fn read_data_str(offset: &mut isize) -> &'static str {
+    let data: *const u8 = unsafe { env.host_field_data.offset(*offset) };
+    let byte_len = unsafe { (data as *const u32).read() } as usize;
+    *offset += 4;
+    let data: *const u8 = unsafe { env.host_field_data.offset(*offset) };
+    let str_data = unsafe { std::slice::from_raw_parts(data, byte_len) };
+    *offset += byte_len as isize;
+    let rem = *offset % 4;
+    if rem > 0 {
+        *offset += 4 - rem;
+    }
+    unsafe { core::str::from_utf8_unchecked(str_data) }
+}
+
+impl Environment for VirtAdapter {
+    fn get_environment() -> Vec<(String, String)> {
+        let mut environment = Vec::new();
+        let mut data_offset: isize = 0;
+        for _ in 0..unsafe { env.host_field_cnt } {
+            let env_key = read_data_str(&mut data_offset);
+            let env_val = read_data_str(&mut data_offset);
+            environment.push((env_key.to_string(), env_val.to_string()));
+        }
+        // fallback ASSUMES that all data is alphabetically ordered
+        if unsafe { env.host_fallback } {
+            let mut allow_or_deny = Vec::new();
+            for _ in 0..unsafe { env.host_allow_or_deny_cnt } {
+                let allow_or_deny_key = read_data_str(&mut data_offset);
+                allow_or_deny.push(allow_or_deny_key);
+            }
+
+            let is_allow_list = unsafe { env.host_fallback_allow };
+            for (key, value) in wasi::cli_base::environment::get_environment().drain(..) {
+                let in_list = allow_or_deny.binary_search(&key.as_ref()).is_ok();
+                if is_allow_list && in_list || !is_allow_list && !in_list {
+                    environment.push((key, value));
```

Comment:
What happens if there is an override specified which also has a matching variable supplied by the host (i.e. `host_fallback` is true)?

If it's an allow list and the override variable is in the allow list, it'll get multiple entries in `environment` with the same key, right? Same goes for a deny list and it isn't in the deny list.

Duplicate environment variables have undefined semantics in POSIX (`getenv` is usually implemented as a linear search in libc implementations, so our overrides would happen to "win" in that case); as far as I am aware (but definitely could be wrong), WASI does not define duplicate semantics nor require an ordering so that it behaves similarly to POSIX's `environ`.

Perhaps we should just build a `HashMap` (or `BTreeMap`) and insert the overrides after we've inserted what we should from the host (if applicable)?

At that point, perhaps we should cache it with a [OnceLock](https://doc.rust-lang.org/stable/std/sync/struct.OnceLock.html)?

Reply:
Thanks, I meant to do a binary search over the overrides to exclude them always from the host pass through, but forgot to implement that. And this is exactly why `overrides` is supposed to be ordered. We could potentially have a new customization to overrides where there are both `env_overrides` and `env_defaults` but perhaps that can be a separate feature rather.

Reply:
Pushed up an adjustment, thanks for bringing this up.

Reply:
Ok, looks good now!